feat: Add developer option for find and replace

lukstbit opened the pull request:
## Purpose / Description

Enables desktop's find and replace menu option(implemented as a dialog). The line count is big but it's mostly tests and code verbosity, the actual feature is small. Some notes:

- decided against a viewmodel in the new dialog fragment, we should probably make these dialogs as simple as possible
- added some tests, the tests related to the fragment's ui should probably be on their own but it would require extra work to make the activity viewmodel replaceable(and I didn't see it worth it to modify the production code for easier testing)
- browser also allows find and replace on a field directly, see documentation on the new dialog fragment

How it looks:


https://github.com/user-attachments/assets/988fee01-5e54-4133-abdf-680ba4e6ca60


<img width="300" height="700" src="https://github.com/user-attachments/assets/95cf69e4-1ce5-4937-b0b4-5a4c30c31a4b" /><img width="300" height="700" src="https://github.com/user-attachments/assets/42bcb0bd-97b1-4565-998e-ebfb2f52c146"/><img width="300" height="700" src="https://github.com/user-attachments/assets/55b03cf6-31ae-42c5-8989-15fbcea4ff7a"/>
![landscape](https://github.com/user-attachments/assets/dc22f9d8-2217-4fdb-a625-485f5fe2de69)

![find_replace_desktop](https://github.com/user-attachments/assets/61ae5a0e-028b-4779-9aa9-4d160c9f1e70)


## Fixes
* Fixes #16386 

## How Has This Been Tested?

Manually tested the new dialog in various scenarios. Added some tests. Ran all tests.

## Checklist

- [x] You have a descriptive commit message with a short title (first line, max 50 chars).
- [x] You have commented your code, particularly in hard-to-understand areas
- [x] You have performed a self-review of your own code
- [x] UI changes: include screenshots of all affected screens (in particular showing any new or changed strings)
- [ ] UI Changes: You have tested your change using the [Google Accessibility Scanner](https://play.google.com/store/apps/details?id=com.google.android.apps.accessibility.auditor)

david-allison commented on AnkiDroid/src/main/java/com/ichi2/anki/CardBrowser.kt:
```diff
@@ -456,6 +467,37 @@ open class CardBrowser :
+                withProgress {
+                    // TODO pass the selection as the user saw it in the dialog to avoid running "find
+                    //  and replace" on a different selection
+                    val noteIds =
```
Move to the ViewModel

```patch
Subject: [PATCH] NF: integration test - Filtered Deck Options

Initial implementation by David 

This test ensures that every changes possibly done in the deck options
are correctly applied

Co-authored-by: David Allison <62114487+david-allison@users.noreply.github.com>
---
Index: AnkiDroid/src/main/java/com/ichi2/anki/browser/CardBrowserViewModel.kt
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/AnkiDroid/src/main/java/com/ichi2/anki/browser/CardBrowserViewModel.kt b/AnkiDroid/src/main/java/com/ichi2/anki/browser/CardBrowserViewModel.kt
--- a/AnkiDroid/src/main/java/com/ichi2/anki/browser/CardBrowserViewModel.kt	(revision 3e52c0792f8f1a58737ba104676cd4ce39dd9c1b)
+++ b/AnkiDroid/src/main/java/com/ichi2/anki/browser/CardBrowserViewModel.kt	(date 1739552134619)
@@ -37,6 +37,8 @@
 import com.ichi2.anki.DeckSpinnerSelection.Companion.ALL_DECKS_ID
 import com.ichi2.anki.Flag
 import com.ichi2.anki.PreviewerDestination
+import com.ichi2.anki.browser.FindAndReplaceDialogFragment.Companion.ALL_FIELDS_AS_FIELD
+import com.ichi2.anki.browser.FindAndReplaceDialogFragment.Companion.TAGS_AS_FIELD
 import com.ichi2.anki.browser.RepositionCardsRequest.RepositionData
 import com.ichi2.anki.export.ExportDialogFragment.ExportType
 import com.ichi2.anki.launchCatchingIO
@@ -53,10 +55,12 @@
 import com.ichi2.libanki.CardId
 import com.ichi2.libanki.CardType
 import com.ichi2.libanki.ChangeManager
+import com.ichi2.libanki.Collection
 import com.ichi2.libanki.DeckId
 import com.ichi2.libanki.QueueType
 import com.ichi2.libanki.QueueType.ManuallyBuried
 import com.ichi2.libanki.QueueType.SiblingBuried
+import com.ichi2.libanki.Tags
 import com.ichi2.libanki.undoableOp
 import kotlinx.coroutines.Deferred
 import kotlinx.coroutines.Job
@@ -998,6 +1002,38 @@
 
     private fun refreshSearch() = launchSearchForCards()
 
+    /**
+     * Replace occurrences of a search with a new value
+     * 
+     * @see Collection.findReplace
+     * @see Tags.findAndReplace
+     * @return The number of affected notes
+     */
+    fun findAndReplace(
+        targetField: String,
+        search: String,
+        replacement: String,
+        onlyOnSelectedNotes: Boolean,
+        matchCase: Boolean,
+        regex: Boolean
+    ) = viewModelScope.async {
+        // TODO pass the selection as the user saw it in the dialog to avoid running "find
+        //  and replace" on a different selection
+        val noteIds =  if (onlyOnSelectedNotes) queryAllSelectedNoteIds() else emptyList()
+
+        if (targetField == TAGS_AS_FIELD) {
+            undoableOp {
+                tags.findAndReplace(noteIds, search, replacement, regex, matchCase)
+            }.count
+        } else {
+            val field =
+                if (targetField == ALL_FIELDS_AS_FIELD) null else targetField
+            undoableOp {
+                findReplace(noteIds, search, replacement, regex, field, matchCase)
+            }.count
+        }
+    }
+
     private suspend fun clearCardsList() {
         cards.reset()
         flowOfCardsUpdated.emit(Unit)
Index: AnkiDroid/src/main/java/com/ichi2/libanki/Collection.kt
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/AnkiDroid/src/main/java/com/ichi2/libanki/Collection.kt b/AnkiDroid/src/main/java/com/ichi2/libanki/Collection.kt
--- a/AnkiDroid/src/main/java/com/ichi2/libanki/Collection.kt	(revision 3e52c0792f8f1a58737ba104676cd4ce39dd9c1b)
+++ b/AnkiDroid/src/main/java/com/ichi2/libanki/Collection.kt	(date 1739552284756)
@@ -395,6 +395,11 @@
         return noteIDsList
     }
 
+
+    /**
+     * @return An [OpChangesWithCount] representing the number of affected notes
+     */
+    @CheckResult
     @LibAnkiAlias("find_and_replace")
     fun findReplace(
         nids: List<Long>,
Index: AnkiDroid/src/main/java/com/ichi2/anki/CardBrowser.kt
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/AnkiDroid/src/main/java/com/ichi2/anki/CardBrowser.kt b/AnkiDroid/src/main/java/com/ichi2/anki/CardBrowser.kt
--- a/AnkiDroid/src/main/java/com/ichi2/anki/CardBrowser.kt	(revision 3e52c0792f8f1a58737ba104676cd4ce39dd9c1b)
+++ b/AnkiDroid/src/main/java/com/ichi2/anki/CardBrowser.kt	(date 1739552134612)
@@ -68,7 +68,6 @@
 import com.ichi2.anki.browser.CardOrNoteId
 import com.ichi2.anki.browser.ColumnHeading
 import com.ichi2.anki.browser.FindAndReplaceDialogFragment
-import com.ichi2.anki.browser.FindAndReplaceDialogFragment.Companion.ALL_FIELDS_AS_FIELD
 import com.ichi2.anki.browser.FindAndReplaceDialogFragment.Companion.ARG_FIELD
 import com.ichi2.anki.browser.FindAndReplaceDialogFragment.Companion.ARG_MATCH_CASE
 import com.ichi2.anki.browser.FindAndReplaceDialogFragment.Companion.ARG_ONLY_SELECTED_NOTES
@@ -76,7 +75,6 @@
 import com.ichi2.anki.browser.FindAndReplaceDialogFragment.Companion.ARG_REPLACEMENT
 import com.ichi2.anki.browser.FindAndReplaceDialogFragment.Companion.ARG_SEARCH
 import com.ichi2.anki.browser.FindAndReplaceDialogFragment.Companion.REQUEST_FIND_AND_REPLACE
-import com.ichi2.anki.browser.FindAndReplaceDialogFragment.Companion.TAGS_AS_FIELD
 import com.ichi2.anki.browser.PreviewerIdsFile
 import com.ichi2.anki.browser.RepositionCardFragment
 import com.ichi2.anki.browser.RepositionCardFragment.Companion.REQUEST_REPOSITION_NEW_CARDS
@@ -476,26 +474,17 @@
                 val onlyOnSelectedNotes = bundle.getBoolean(ARG_ONLY_SELECTED_NOTES, true)
                 val matchCase = bundle.getBoolean(ARG_MATCH_CASE, false)
                 val regex = bundle.getBoolean(ARG_REGEX, false)
-                withProgress {
-                    // TODO pass the selection as the user saw it in the dialog to avoid running "find
-                    //  and replace" on a different selection
-                    val noteIds =
-                        if (onlyOnSelectedNotes) viewModel.queryAllSelectedNoteIds() else emptyList()
-
-                    val count =
-                        if (targetField == TAGS_AS_FIELD) {
-                            undoableOp {
-                                tags.findAndReplace(noteIds, search, replacement, regex, matchCase)
-                            }.count
-                        } else {
-                            val field =
-                                if (targetField == ALL_FIELDS_AS_FIELD) null else targetField
-                            undoableOp {
-                                findReplace(noteIds, search, replacement, regex, field, matchCase)
-                            }.count
-                        }
-                    showSnackbar(TR.browsingNotesUpdated(count))
-                }
+                val count = withProgress {
+                    viewModel.findAndReplace(
+                        targetField,
+                        search,
+                        replacement,
+                        onlyOnSelectedNotes,
+                        matchCase,
+                        regex
+                    ).await()
+                }
+                showSnackbar(TR.browsingNotesUpdated(count))
             }
         }
     }
Index: AnkiDroid/src/main/java/com/ichi2/libanki/Tags.kt
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/AnkiDroid/src/main/java/com/ichi2/libanki/Tags.kt b/AnkiDroid/src/main/java/com/ichi2/libanki/Tags.kt
--- a/AnkiDroid/src/main/java/com/ichi2/libanki/Tags.kt	(revision 3e52c0792f8f1a58737ba104676cd4ce39dd9c1b)
+++ b/AnkiDroid/src/main/java/com/ichi2/libanki/Tags.kt	(date 1739552387220)
@@ -17,6 +17,7 @@
  ****************************************************************************************/
 package com.ichi2.libanki
 
+import androidx.annotation.CheckResult
 import androidx.annotation.WorkerThread
 import anki.collection.OpChangesWithCount
 import com.ichi2.libanki.utils.LibAnkiAlias
@@ -113,7 +114,15 @@
         tags: Iterable<String>,
     ): Boolean = tags.map { it.lowercase() }.contains(tag.lowercase())
 
+
+
+    /**
+     * Replace occurrences of a search with a new value in tags.
+     * 
+     * @return An [OpChangesWithCount] representing the number of affected notes
+     */
     // https://github.com/ankitects/anki/blob/main/pylib/anki/tags.py#L73
+    @CheckResult
     @LibAnkiAlias("find_and_replace")
     fun findAndReplace(
         noteIds: List<Long>,
```

david-allison reviewed (approved):
LGTM, since it's a dev option, all are optional

lukstbit commented:
I implemented all the review requests.

I also changed the UI a bit to align the title and content start:

![image](https://github.com/user-attachments/assets/fa687713-2707-4e77-8ba4-f09ca1906d2e)

david-allison commented:
Visually: still an approve. Comments:

* I'd use Material Text Fields for the first 2 fields
* Reduce the left padding on the checkboxes, to align them with the spinner

david-allison commented on AnkiDroid/src/main/java/com/ichi2/anki/CardBrowser.kt:
```diff
@@ -443,6 +454,37 @@ open class CardBrowser :
+
+        supportFragmentManager.setFragmentResultListener(REQUEST_FIND_AND_REPLACE, this) { _, bundle ->
+            launchCatchingTask {
+                val targetField = bundle.getString(ARG_FIELD) ?: return@launchCatchingTask
+                val search = bundle.getString(ARG_SEARCH) ?: return@launchCatchingTask
+                val replacement = bundle.getString(ARG_REPLACEMENT) ?: return@launchCatchingTask
+                val onlyOnSelectedNotes = bundle.getBoolean(ARG_ONLY_SELECTED_NOTES, true)
+                val matchCase = bundle.getBoolean(ARG_MATCH_CASE, false)
+                val regex = bundle.getBoolean(ARG_REGEX, false)
```
Nothing we can do similar to `@Parcelize`?

Don't let me block you here, if it's not a generic solution, totally up to you: I wouldn't ask for it, but probably usable now it's released

lukstbit replied:
> Nothing we can do similar to @Parcelize?

I'm not sure, I'll have a look this week.

lukstbit commented:
> I'd use Material Text Fields for the first 2 fields
>Reduce the left padding on the checkboxes, to align them with the spinner

Done. Didn't use the above labels as hints, I'm not sure they are a match and the current design it's just good enough.

> I personally prefer using the "?" in the title, rather than a 'help' neutral button

I don't think a ? it's appropriate in the title. Also, I value the help button and I think it's going to be useful especially for users that want to try some regex(thanks to the links in the docs).
![image](https://github.com/user-attachments/assets/ded06233-5e3c-4bde-ad14-21b96a86b78b)

david-allison reviewed (approved):
I'm not here to naysay, and it's a developer option. Huge improvement!

EDIT: Unit test failure

lukstbit commented:
Fixed the unit tests and also applied the patch!

BrayanDSO reviewed (approved):
Looks good. Nothing block for a dev option
